GET_DEVICES improve, update plan, use normative

bertrand-marquis opened the pull request:
In this pull request, I try to do the following:
- improve GET_DEVICES message description as it was using number in an imprecise way
- sync the plan to other transport definitions (remove introduction, remove BUS operations, have discovery, initialization and operation to mimic other transports)
- start to switch to normative statements (only done for now in basic concept and discovery)

Would be nice to have some feedback on the normative and validate the new plan before i finish this.

PDF version of this PR as of 2025-12-03 5PM UTC:
https://github.com/wmamills/virtio-msg-spec/releases/download/bertrand-pr28-wmamills-draft-v31/bertrand-pr28-wmamills-draft-v31.pdf

bertrand-marquis committed:
virtio-msg: rework GET_DEVICES

Rework GET_DEVICES message description to prevent confusion between
Number of devices and device number by using device-number for the
device number identifier and renaming fields of the message.
Also provide an example to illustrate the usage of the message.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: remove emph or textbf for MUST/MAY/SHOULD

To comply with the rest of the OASIS Virtio specification removing all
emph or textbf around MUST/MAY/SHOULD.
Do some minor rewording to prevent using normative wording when not
required.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: Remove introduction as chapter

To follow the organization of other transports, remove the introduction
chapter by moving its content into basic concepts and just having a
small introduction paragraph for the chapter instead.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: create a Device discovery section

Create a stand-alone device discovery section to have the same
organization as other transports.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: remove bus operation section

Merge content into messages description directly by adding an intended
usage paragraph or into the Basic Concepts section.

This makes the transport content look more like PCI or Channel IO.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: merge all bus vs transport chapters in one

Rework the Basic concepts to merge all chapters speaking of bus vs
transport and the bus messages and transport messages into a single one
to prevent repetition.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: move compatibility rules in basic concepts

Move version compatibility rules in a sub-chapter of transport revision
in the basic concepts.
This is to prepare for the removal of the compliance chapter.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: add new command for bus normative

Add a new command to have a standard way to define virtio-msg bus
normative statements.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: normative in bus vs transport relationship

Add bus normative statement in bus vs transport section of basic
concepts.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: rework transport revision

Rework the transport revision and maximum message size section to switch
to normative declarations.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: Switch to normative in device numbers

Switch to normative statement in device numbers and enumeration section.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: switch generation count to normative

Rework Generation Count section in basic concept to use normative
statements.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: apply normative in feature blocks

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: Switch error signaling to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: switch endianness to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert common message format to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert correlation to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert device discovery to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert flow overview to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert device information to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert feature negotiation to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: switch device configuration to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: Switch virqueue config to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: switch status information to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: switch finalize init to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert driver notifications to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: switch device notifications to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert config changes to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert virtqueue change to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert reset and shutdown to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert hotplug and removal to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert transport messages intro to normative

Also convert the overview

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert GET_DEVICE_INFO to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert get_features to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert set_features to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert get/set config to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert get/set status to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert vqueue messages to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert get_shm to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert event messages to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: reword bus message intro and overview

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert get_devices to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: convert event dev and ping to normative

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: remove compliance chapter

All the content of the chapter is now present in normative statements in
the required sections.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: Simplify correlation handling

Rework the correlation handling:
- fully delegate correlation to the bus
- only enforce for answers to be received in the order of the requests
- remove restriction to set token to 0 for events
- enforce for the device to not make any assumption on the token value

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: Remove (xx) in normative

Remove (Driver), (Device) or (Bus) as those are repeating the
requirement title.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: add conformance entries in conformance chapter

Add entries for conformance of Virtio over Message for Driver, Device
and bus.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

epilys commented on transport-msg.tex:
```diff
@@ -71,12 +77,61 @@ \subsubsection{High-Level Architecture}
+    These messages are \emph{optional} in environments where
+    device discovery or state changes occur through other means (e.g., device
+    tree). However, if a bus chooses to handle those tasks via messages,
+    it SHOULD implement the appropriate bus message definitions.
```
Suggestion: move `SHOULD` statement to busnormative section below

bertrand-marquis replied:
Agree

bertrand-marquis replied:
Done

epilys commented on transport-msg.tex:
```diff
@@ -71,12 +77,61 @@ \subsubsection{High-Level Architecture}
+This separation lets a bus remain minimal if it obtains device information from
+firmware tables, while still supporting fully message-based enumeration and
+hotplug when desired.
+
```
Change order:

```suggestion

This separation lets a bus remain minimal if it obtains device information from
firmware tables, while still supporting fully message-based enumeration and
hotplug when desired.

\busnormative{\paragraph}{Transport Message Forwarding}{Virtio Transport Options / Virtio Over Messages / Basic Concepts / Relationship between bus and transport / Transport Message Forwarding}
\begin{itemize}
  \item A bus implementation MUST relay each transport message to the device
        number identified in the message header, regardless of how it
        discovered or enumerated that device.
  \item A bus implementation SHOULD treat transport messages as opaque apart
        from enforcing generic transport limits, such as the advertised maximum
        message size, and SHOULD NOT modify the transport payload.
\end{itemize}

```

bertrand-marquis replied:
Agree

bertrand-marquis replied:
Done

epilys commented on transport-msg.tex:
```diff
@@ -153,59 +207,82 @@ \subsubsection{Transport Revisions and Maximum Message Size}
+imposed by the underlying bus (for example, a framing header) does not count
+against this limit. Today the maximum useful transport payload is expected to
+be 274 bytes, which is sufficient for a GET_CONFIG or SET_CONFIG transfer with
+256 bytes of configuration data.
```
> Today the maximum useful transport payload is expected to
be 274 bytes, which is sufficient for a GET_CONFIG or SET_CONFIG transfer with
256 bytes of configuration data

Not ideal phrasing here. How about turning this into a `SHOULD` normative statement?

bertrand-marquis replied:
I am a bit wondering if we should not remove this size statement all together as this is really a finger in the air statement, current PoCs are using size smaller than that and ffa is using 104 bytes.

bertrand-marquis replied:
Did a try, let me know what you think

epilys commented on transport-msg.tex:
```diff
@@ -153,59 +207,82 @@ \subsubsection{Transport Revisions and Maximum Message Size}
+revision still implement the mandatory messages and semantics defined in prior
+revisions. When a bus instance, driver, or device encounters an unsupported
+message or feature, it should reject the request explicitly or ignore it rather
+than exhibit undefined behavior.
```
This sounds like potential normative statements, too

bertrand-marquis replied:
i will give that a try

bertrand-marquis replied:
Did a try, let me know what you think

epilys commented on transport-msg.tex:
```diff
@@ -214,97 +291,81 @@ \subsubsection{Configuration Generation Count}
 
-A device \emph{MUST} respond with all zero feature data for features bit
-requested beyond those the device implements.
+\devicenormative{\paragraph}{Feature Blocks (Device)}{Virtio Transport Options / Virtio Over Messages / Basic Concepts / FeatureBlocks / Device}
```
Is `FeatureBlocks` instead of `Feature Blocks` on purpose?

bertrand-marquis replied:
This is reusing the label format upper but happy to put full naming here and elsewhere for coherency

bertrand-marquis replied:
done

epilys commented on transport-msg.tex:
```diff
@@ -1240,33 +1323,38 @@ \subsection{Bus Messages}\label{sec:Virtio Transport Options / Virtio Over Messa
-and \textbf{MAY} in this document are to be interpreted as specified by
+The terms MUST, MUST NOT, SHOULD, SHOULD NOT,
+and MAY in this document are to be interpreted as specified by
 \href{https://www.rfc-editor.org/rfc/rfc2119}{RFC 2119}.
```
this is not necessary, since it's stated in 1.3 Terminology of the spec

bertrand-marquis replied:
The complete compliance section will be removed once we have switch to normative

bertrand-marquis replied:
compliance is now removed

bertrand-marquis committed:
virtio-msg: Improve message ID specification

Use a table to properly define the message IDs area reserved for
specific usage.
Use the same format and fully specify also the message IDs area for bus
messages.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: Use struct instead of tables for messages

Define the messages format using structures instead of using table to be
more coherent with other transports and the rest of virtio
specification.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: Add normative statement for message size

Add normative statement on message size constraints in the revisions
chapter.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: Add normative statements for version compat

Define some normative statements for backward compatibility rules.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: Remove compressed chapter lable/normative

Replace compressed forms of labels or normative with long forms (for
example FeatureBlocks to Feature Blocks).

While there update conformance chapter.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
virtio-msg: move ping before event_device

To keep messages in the order of the table of IDs, move ping definition
before event_device.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis commented:
I updated the pull request with the following:
- full normative rework including adding entries in conformance.tex
- switch message definitions to use struct instead of tables: tell me if you think this is better, easy to revert
- moved ping before event_device to have the right order
- rework correlation chapter to simplify as said during last meeting
- clear up message ID numbers by having numbers defined in tables to have clear range for msg/event/custom msg/event on both transport and bus messages

wmamills reviewed (commented):
This is ready for the PATCH V1

wmamills reviewed (approved):
I will approve and merge to the virtio-msg-rfc2 branch and then squash and prepare virtio-msg-patch1 branch
